### PR TITLE
refact: CaseData model to be used in callback response (CMC-849)

### PR DESCRIPTION
**Pull request by sabahirfan**
### JIRA link (if applicable) ###
[CMC-849](url)

### Change description ###

The callback handler uses the `CaseData` model to be returned back via its `CallbackResponse`

This PR also removes the duplicate validation logic from the `BusinessProcessService`. As the same is already being invoked in `NoOngoingBusinessProcessAspect` 

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/unspec/service/BusinessProcessService.java [hunk 1]
```diff
@@ -1,31 +1,26 @@
...
 @RequiredArgsConstructor
 public class BusinessProcessService {
 
-    private static final String ERROR_MESSAGE = "Business Process Error";
-
-    private final ObjectMapper mapper;
-
-    public List<String> updateBusinessProcess(Map<String, Object> data, CaseEvent caseEvent) {
-        CaseData caseData = mapper.convertValue(data, CaseData.class);
+    public CaseData updateBusinessProcess(CaseData caseData, CaseEvent caseEvent) {
```
Comment (pd-dudkiewicz) [inline src/main/java/uk/gov/hmcts/reform/unspec/service/BusinessProcessService.java:15]: We still need to pass data map here, otherwise business process will not be set to READY

Comment (sabahirfan) [reply src/main/java/uk/gov/hmcts/reform/unspec/service/BusinessProcessService.java:15]: good catch. done

File: src/main/java/uk/gov/hmcts/reform/unspec/aspect/EventEmitterAspect.java [hunk 2]
```diff
@@ -20,15 +19,14 @@
...
     private final EventEmitterService eventEmitterService;
 
     @Around("execution(* *(*)) && @annotation(EventEmitter) && args(callbackParams))")
     public Object emitBusinessProcessEvent(ProceedingJoinPoint joinPoint, CallbackParams callbackParams)
         throws Throwable {
         if (callbackParams.getType() == SUBMITTED) {
-            CaseData caseData = caseDetailsConverter.toCaseData(callbackParams.getRequest().getCaseDetails());
-            if (caseData.getBusinessProcess().getStatus() == READY) {
+            CaseData caseData = callbackParams.getCaseData();
+            if (caseData.getBusinessProcess() != null && caseData.getBusinessProcess().getStatus() == READY) {
```
Comment (swalker125) [inline src/main/java/uk/gov/hmcts/reform/unspec/aspect/EventEmitterAspect.java:29]: would prefer this logic contained in BusinessProcess but not really a part of your PR so fine to leave I guess!

File: src/main/java/uk/gov/hmcts/reform/unspec/handler/callback/AcknowledgeServiceCallbackHandler.java [hunk 3]
```diff
@@ -66,17 +68,17 @@ private CallbackResponse validateDateOfBirth(CallbackParams callbackParams) {
...
 
-        data.put(RESPONSE_DEADLINE, newResponseDate.atTime(MID_NIGHT));
-        List<String> errors = businessProcessService.updateBusinessProcess(data, ACKNOWLEDGE_SERVICE);
+        CaseData caseDataUpdated = businessProcessService.updateBusinessProcess(
+            caseDataBuilder.build(), ACKNOWLEDGE_SERVICE);
 
         return AboutToStartOrSubmitCallbackResponse.builder()
-            .data(data)
-            .errors(errors)
+            .data(caseDetailsConverter.toMap(caseDataUpdated))
```
Comment (swalker125) [inline src/main/java/uk/gov/hmcts/reform/unspec/handler/callback/AcknowledgeServiceCallbackHandler.java:81]: what about having a toMap method on CaseData?

File: src/main/java/uk/gov/hmcts/reform/unspec/handler/callback/AcknowledgeServiceCallbackHandler.java [hunk 4]
```diff
@@ -66,17 +68,17 @@ private CallbackResponse validateDateOfBirth(CallbackParams callbackParams) {
...
-
+        CaseData caseData = callbackParams.getCaseData();
+        LocalDateTime responseDeadline = caseData.getRespondentSolicitor1ResponseDeadline();
         LocalDate newResponseDate = workingDayIndicator.getNextWorkingDay(responseDeadline.plusDays(14).toLocalDate());
+        CaseData.CaseDataBuilder caseDataBuilder = caseData.toBuilder()
+            .respondentSolicitor1ResponseDeadline(newResponseDate.atTime(MID_NIGHT));
 
-        data.put(RESPONSE_DEADLINE, newResponseDate.atTime(MID_NIGHT));
-        List<String> errors = businessProcessService.updateBusinessProcess(data, ACKNOWLEDGE_SERVICE);
+        CaseData caseDataUpdated = businessProcessService.updateBusinessProcess(
```
Comment (swalker125) [inline src/main/java/uk/gov/hmcts/reform/unspec/handler/callback/AcknowledgeServiceCallbackHandler.java:77]: updateBusinessProcess returning CaseData seems a lot? I would have thought this would be nicer as a void method on BusinessProcess? I'm not actually sure of a use for BusinessProcessService. Again maybe not this PRs concern

File: src/main/java/uk/gov/hmcts/reform/unspec/handler/callback/ConfirmServiceCallbackHandler.java [hunk 5]
```diff
@@ -78,11 +80,8 @@
     private CallbackResponse prepopulateServedDocuments(CallbackParams callbackParams) {
         List<ServedDocuments> servedDocuments = List.of(ServedDocuments.CLAIM_FORM);
 
-        Map<String, Object> data = callbackParams.getRequest().getCaseDetails().getData();
-        data.put("servedDocuments", servedDocuments);
-
         return AboutToStartOrSubmitCallbackResponse.builder()
-            .data(data)
+            .data(Map.of("servedDocuments", servedDocuments))
```
Comment (swalker125) [inline src/main/java/uk/gov/hmcts/reform/unspec/handler/callback/ConfirmServiceCallbackHandler.java]: is the point of this PR not to remove bits like this?

File: src/main/java/uk/gov/hmcts/reform/unspec/handler/callback/ConfirmServiceCallbackHandler.java [hunk 6]
```diff
@@ -119,34 +118,29 @@ private CallbackResponse prepareCertificateOfService(CallbackParams callbackPara
...
-        CaseData caseDateUpdated = caseData.toBuilder()
+        CaseData.CaseDataBuilder caseDataBuilder = caseData.toBuilder()
             .deemedServiceDateToRespondentSolicitor1(deemedDateOfService)
-            .respondentSolicitor1ResponseDeadline(responseDeadline)
-            .build();
+            .respondentSolicitor1ResponseDeadline(responseDeadline);
 
         CaseDocument certificateOfService = certificateOfServiceGenerator.generate(
-            caseDateUpdated,
+            caseDataBuilder.build(),
```
Comment (swalker125) [inline src/main/java/uk/gov/hmcts/reform/unspec/handler/callback/ConfirmServiceCallbackHandler.java:132]: really don't like the requirement to build case data here, is hard to know what is needed for generate method. Again - maybe not your PRs concern 😅

File: src/main/java/uk/gov/hmcts/reform/unspec/handler/callback/DispatchBusinessProcessCallbackHandler.java [hunk 7]
```diff
@@ -36,17 +40,19 @@
...
         if (businessProcess.getStatus() == READY) {
-            data.put("businessProcess", BusinessProcess.builder()
-                .camundaEvent(businessProcess.getCamundaEvent())
-                .status(DISPATCHED)
-                .build());
+            caseDataBuilder
+                .businessProcess(BusinessProcess.builder()
+                                     .camundaEvent(businessProcess.getCamundaEvent())
+                                     .status(DISPATCHED)
+                                     .build());
```
Comment (swalker125) [inline src/main/java/uk/gov/hmcts/reform/unspec/handler/callback/DispatchBusinessProcessCallbackHandler.java:51]: good example of why businessProcessService is now not needed

Comment (swalker125) [reply src/main/java/uk/gov/hmcts/reform/unspec/handler/callback/DispatchBusinessProcessCallbackHandler.java:51]: also another use case for some void methods to update business process in case data directly rather than need case data builder

File: src/main/java/uk/gov/hmcts/reform/unspec/model/CaseData.java [hunk 8]
```diff
@@ -30,7 +31,9 @@
 @HasServiceDateTheSameAsOrAfterIssueDate(groups = ConfirmServiceDateGroup.class)
 public class CaseData {
 
+    @JsonProperty(access = JsonProperty.Access.WRITE_ONLY)
     private final Long ccdCaseReference;
+    @JsonProperty(access = JsonProperty.Access.WRITE_ONLY)
     private final CaseState ccdState;
```
Comment (swalker125) [inline src/main/java/uk/gov/hmcts/reform/unspec/model/CaseData.java:37]: why are these added to CaseData?

File: src/test/java/uk/gov/hmcts/reform/unspec/handler/callback/DispatchBusinessProcessCallbackHandlerTest.java [hunk 9]
```diff
@@ -32,23 +36,33 @@
...
+                .extracting("camundaEvent")
+                .isEqualTo("testCamundaEvent");
+            assertThat(response.getData())
+                .extracting("businessProcess")
+                .extracting("activityId")
+                .isNull();
+            assertThat(response.getData())
+                .extracting("businessProcess")
+                .extracting("processInstanceId")
+                .isNull();
```
Comment (swalker125) [inline src/test/java/uk/gov/hmcts/reform/unspec/handler/callback/DispatchBusinessProcessCallbackHandlerTest.java]: does .extracting not take multiple strings for assertions / can you not chain from .extracting business process

File: src/test/java/uk/gov/hmcts/reform/unspec/handler/callback/RequestExtensionCallbackHandlerTest.java [hunk 10]
```diff
@@ -155,7 +141,7 @@ void shouldUpdateResponseDeadlineToProposedDeadline_whenExtensionAlreadyAgreed()
             var response = (AboutToStartOrSubmitCallbackResponse) handler.handle(params);
 
             assertThat(response.getData())
-                .containsEntry(RESPONSE_DEADLINE, proposedDeadline.atTime(MID_NIGHT));
+                .containsEntry(RESPONSE_DEADLINE, proposedDeadline.atTime(MID_NIGHT).toString());
```
Comment (swalker125) [inline src/test/java/uk/gov/hmcts/reform/unspec/handler/callback/RequestExtensionCallbackHandlerTest.java:144]: suspicious toString here

File: src/test/java/uk/gov/hmcts/reform/unspec/handler/callback/RequestExtensionCallbackHandlerTest.java [hunk 11]
```diff
@@ -178,9 +164,18 @@ void shouldNotUpdateResponseDeadline_whenExtensionIsNotAlreadyAgreed() {
...
-            verify(businessProcessService).updateBusinessProcess(new HashMap<>(), REQUEST_EXTENSION);
+            assertThat(response.getData())
+                .extracting("businessProcess")
+                .extracting("camundaEvent")
+                .isEqualTo(REQUEST_EXTENSION.name());
+
+            assertThat(response.getData())
+                .extracting("businessProcess")
+                .extracting("status")
+                .isEqualTo("READY");
```
Comment (swalker125) [inline src/test/java/uk/gov/hmcts/reform/unspec/handler/callback/RequestExtensionCallbackHandlerTest.java]: again I think you can extract multiple / chain here

**Review by swalker125 — COMMENTED** [review]
Couple of things, maybe not a part of your PR though.

- I don't see point of BusinessProcessService and it's needed in a few places
- I think case data should have a toMap method rather than requiring CaseDetailsConvertor everywhere. Whether this is a method in CaseData or inherited from another class I don't know. What do you reckon? We could have a base type like DocmosisData that has a toMap method
- Why aren't we just using setters directly on CaseData, is it cleaner than using toBuilder everywhere?

File: src/main/java/uk/gov/hmcts/reform/unspec/handler/callback/AcknowledgeServiceCallbackHandler.java (hunk 4 above)
Comment (sabahirfan) [reply src/main/java/uk/gov/hmcts/reform/unspec/handler/callback/AcknowledgeServiceCallbackHandler.java:77]: I agree that `BusinessProcessService` seems unnecessary and we can clean that up as separate PR not part of this.

File: src/main/java/uk/gov/hmcts/reform/unspec/handler/callback/AcknowledgeServiceCallbackHandler.java (hunk 3 above)
Comment (sabahirfan) [reply src/main/java/uk/gov/hmcts/reform/unspec/handler/callback/AcknowledgeServiceCallbackHandler.java:81]: I will prefer the domain objects plain java objects, don't want to inject `ObjectMapper` bean there.

File: src/main/java/uk/gov/hmcts/reform/unspec/handler/callback/ConfirmServiceCallbackHandler.java (hunk 6 above)
Comment (sabahirfan) [reply src/main/java/uk/gov/hmcts/reform/unspec/handler/callback/ConfirmServiceCallbackHandler.java:132]: will visit this later if needed.

File: src/test/java/uk/gov/hmcts/reform/unspec/handler/callback/RequestExtensionCallbackHandlerTest.java (hunk 10 above)
Comment (sabahirfan) [reply src/test/java/uk/gov/hmcts/reform/unspec/handler/callback/RequestExtensionCallbackHandlerTest.java:144]: assertion fails as in `response.getData()` map it's type is String and we have `LocalDateTime` object, so have to make both of same type.

File: src/test/java/uk/gov/hmcts/reform/unspec/handler/callback/DispatchBusinessProcessCallbackHandlerTest.java (hunk 9 above)
Comment (sabahirfan) [reply src/test/java/uk/gov/hmcts/reform/unspec/handler/callback/DispatchBusinessProcessCallbackHandlerTest.java:65]: I am afraid not. as we have condition on one property. we can not chain after that.

File: src/main/java/uk/gov/hmcts/reform/unspec/handler/callback/ConfirmServiceCallbackHandler.java (hunk 5 above)
Comment (sabahirfan) [reply src/main/java/uk/gov/hmcts/reform/unspec/handler/callback/ConfirmServiceCallbackHandler.java]: This is mid callback where we perpopulate the one field for the particular page. And this is the only use case for this kind of functionality. Sending the complete caseData doesn't seem to be right in this scenario.

File: src/main/java/uk/gov/hmcts/reform/unspec/model/CaseData.java (hunk 8 above)
Comment (sabahirfan) [reply src/main/java/uk/gov/hmcts/reform/unspec/model/CaseData.java:37]: These fields are being used already and we need them in our business logic. But we don't want to save these as part of data map in CaseDetails as we don't have any case field for these.

File: src/test/java/uk/gov/hmcts/reform/unspec/handler/callback/RequestExtensionCallbackHandlerTest.java (hunk 11 above)
Comment (sabahirfan) [reply src/test/java/uk/gov/hmcts/reform/unspec/handler/callback/RequestExtensionCallbackHandlerTest.java:178]: I am afraid not. as we have condition on one property. we can not chain after that.

File: src/main/java/uk/gov/hmcts/reform/unspec/aspect/EventEmitterAspect.java (hunk 2 above)
Comment (sabahirfan) [reply src/main/java/uk/gov/hmcts/reform/unspec/aspect/EventEmitterAspect.java:29]: added null safety to the existing condition.

File: src/test/java/uk/gov/hmcts/reform/unspec/handler/callback/DispatchBusinessProcessCallbackHandlerTest.java (hunk 9 above)
Comment (sabahirfan) [reply src/test/java/uk/gov/hmcts/reform/unspec/handler/callback/DispatchBusinessProcessCallbackHandlerTest.java]: done

File: src/test/java/uk/gov/hmcts/reform/unspec/handler/callback/RequestExtensionCallbackHandlerTest.java (hunk 11 above)
Comment (sabahirfan) [reply src/test/java/uk/gov/hmcts/reform/unspec/handler/callback/RequestExtensionCallbackHandlerTest.java]: done

**Review by dharmendrak — DISMISSED** [review]

File: src/main/java/uk/gov/hmcts/reform/unspec/handler/callback/ConfirmServiceCallbackHandler.java (hunk 5 above)
Comment (sabahirfan) [reply src/main/java/uk/gov/hmcts/reform/unspec/handler/callback/ConfirmServiceCallbackHandler.java]: to be consistent across all changed that as well.

**Review by dharmendrak — APPROVED** [review]

File: src/main/java/uk/gov/hmcts/reform/unspec/handler/callback/AcknowledgeServiceCallbackHandler.java (hunk 3 above)
Comment (swalker125) [reply src/main/java/uk/gov/hmcts/reform/unspec/handler/callback/AcknowledgeServiceCallbackHandler.java:81]: hmm we could have this in an interface as we have for DocmosisData however?

**Review by swalker125 — APPROVED** [review]
